use types in the IR instead of vmap

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43743 Update aliasing in tensorexpr fuser
* **#43742 use types in the IR instead of vmap**
* #43636 Add inline call to profiling executor pipeline
* #43635 Use prim::TensorExprGroup interned symbol
* #43634 Update requires grad property

We can remove all prim::profiles, update the values to their specialized profiled types, and then later guard the input graphs based on the input types of the fusion group. After that we remove specialized tensor types from the graph. This gets rid of having to update the vmap and removes all of the profile nodes in fusing.

Differential Revision: [D23385206](https://our.internmc.facebook.com/intern/diff/D23385206)

Commit:
use types in the IR instead of vmap

[ghstack-poisoned]

Commit:
Update on "use types in the IR instead of vmap"




[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/remove_redundant_profiles.cpp:
```diff
@@ -0,0 +1,46 @@
+      RemoveRedundantProfiles(b, db);
+    }
+
+    if (n->kind() != prim::profile ||
```
Should we do the same for `profile_optional`?

eellison replied:
`profile_optional` profiles the definition instead of each use (bc None type is immutable unlike tensor type), so it's not possible that there's a redundant profile_optional

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -129,30 +130,90 @@ class TensorExprFuser {
   TensorExprFuser(std::shared_ptr<Graph> graph, size_t min_group_size)
       : graph_(std::move(graph)), min_group_size_(min_group_size) {}
 
+  // TODO: handle better
```
Nit: I'm not sure such TODO is helpful - could we either add more details or remove it altogether?

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -129,30 +130,90 @@ class TensorExprFuser {
+  void removeProfileNodesAndSpecializeTypes(Block* b) {
+    for (auto it = b->nodes().begin(); it != b->nodes().end(); it++) {
+      if (it->kind() == prim::profile) {
+        GRAPH_DEBUG("TYPEINFO ERASE: %", it->output()->debugName());
```
There is no typeinfo anymore - should we reword the debug statement?

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -129,30 +130,90 @@ class TensorExprFuser {
+      if (it->kind() == prim::profile) {
+        GRAPH_DEBUG("TYPEINFO ERASE: %", it->output()->debugName());
+        it->output()->replaceAllUsesWith(it->input());
+        if (!profiledWithDifferentTypes(it->input())) {
```
Nit: maybe add the else statement with `GRAPH_DEBUG` to make this situation noticeable?

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -452,62 +462,25 @@ class TensorExprFuser {
+      }
+
+      // fusion outputs are already guarded
+      if (input->node()->kind() == prim::Constant ||
```
Thinking aloud: in future it might be worth adding an optimization to find redundant typechecks and remove them.

ZolotukhinM commented on test/jit/test_profiler.py:
```diff
@@ -65,3 +65,21 @@ def test_fuse(a, b):
         # broadcasts occurred, currently expect to see aten::_grad_sum_to_size
         self.assertIsNotNone(optimized_block.findNode("aten::_grad_sum_to_size"))
+
+    def test_specialized_types(self):
```
Shouldn't existing tests be already testing this? I'd prefer to not adding redundancy unless we're truly testing something unique here.

eellison replied:
I'm testing specifically that the specialized types exist for the nodes we expect them too (TypeCheck / TensorExpr) and not the other nodes (e.g. prim::If). I think that's a new set of tests. `test_fuse` could probably be renamed

ZolotukhinM reviewed (approved):
Cool stuff, please ship!

Commit:
Update on "use types in the IR instead of vmap"


We can remove all prim::profiles, update the values to their specialized profiled types, and then later guard the input graphs based on the input types of the fusion group. After that we remove specialized tensor types from the graph. This gets rid of having to update the vmap and removes all of the profile nodes in fusing.


[ghstack-poisoned]

Commit:
Update on "use types in the IR instead of vmap"


We can remove all prim::profiles, update the values to their specialized profiled types, and then later guard the input graphs based on the input types of the fusion group. After that we remove specialized tensor types from the graph. This gets rid of having to update the vmap and removes all of the profile nodes in fusing.

Differential Revision: [D23385206](https://our.internmc.facebook.com/intern/diff/D23385206)

[ghstack-poisoned]

Commit:
Update on "use types in the IR instead of vmap"


We can remove all prim::profiles, update the values to their specialized profiled types, and then later guard the input graphs based on the input types of the fusion group. After that we remove specialized tensor types from the graph. This gets rid of having to update the vmap and removes all of the profile nodes in fusing.

Differential Revision: [D23385206](https://our.internmc.facebook.com/intern/diff/D23385206)

[ghstack-poisoned]
